src/core/transpile.jl

```diff
@@ -2579,3 +2579,143 @@ function are_gates_at_excluded_positions(
+    )::QuantumCircuit
+
+Throws a `DomainError` if an `Instruction` in the `circuit` operates on an excluded
+connection. The excluded connections are specified with the parameter `excluded_connections`
+in certain `AbstractConnectivity` objects. The `circuit` remains unchanged if no error is
+thrown.
+
+# Examples
+```jldoctest
+julia> excluded_positions = Int[];
+
+julia> excluded_connections = [(2, 3)];
+
+julia> connectivity = LineConnectivity(4, excluded_positions, excluded_connections)
+LineConnectivity{4}
+1──2──3──4
+excluded connections: [(2, 3)]
+
+
+julia> transpiler = RejectGatesOnExcludedConnectionsTranspiler(connectivity);
+
+julia> invalid_circuit = QuantumCircuit(
+                      qubit_count = 4,
+                      instructions = [sigma_x(4), control_z(3, 2)],
+                  )
+Quantum Circuit Object:
+   qubit_count: 4 
+   bit_count: 4 
+q[1]:──────────
+               
+q[2]:───────Z──
+            |  
+q[3]:───────*──
+               
+q[4]:──X───────
+               
+
+
+
+julia> transpile(transpiler, invalid_circuit)
+ERROR: DomainError with LineConnectivity{4}
+1──2──3──4
+excluded connections: [(2, 3)]
+:
+the Snowflurry.ControlZ gate on qubits [3, 2] cannot be applied since connection (2, 3) is unavailable
+[...]
+
+julia> valid_circuit = QuantumCircuit(
+                      qubit_count = 4,
+                      instructions = [sigma_x(1), control_z(3, 4)],
+                  )
+Quantum Circuit Object:
+   qubit_count: 4 
+   bit_count: 4 
+q[1]:──X───────
+               
+q[2]:──────────
+               
+q[3]:───────*──
+            |  
+q[4]:───────Z──
+               
+
+
+
+julia> transpiled_circuit = transpile(transpiler, valid_circuit)
+Quantum Circuit Object:
+   qubit_count: 4 
+   bit_count: 4 
+q[1]:──X───────
+               
+q[2]:──────────
+               
+q[3]:───────*──
+            |  
+q[4]:───────Z──
+               
+
+
+```
+"""
+function transpile(
+    transpiler::RejectGatesOnExcludedConnectionsTranspiler,
+    circuit::QuantumCircuit,
+)::QuantumCircuit
+
+    (found_invalid_gates, message) =
+        are_gates_at_excluded_connections(transpiler.connectivity, circuit)
+
+    if found_invalid_gates
+        throw(DomainError(transpiler.connectivity, message))
+    end
+
+    return circuit
+end
+
+function are_gates_at_excluded_connections(
+    connectivity::AllToAllConnectivity,
+    circuit::QuantumCircuit,
+)::Tuple{Bool,String}
+
+    return (false, "")
+end
+
+function are_gates_at_excluded_connections(
+    connectivity::Union{LineConnectivity,LatticeConnectivity},
+    circuit::QuantumCircuit,
+)::Tuple{Bool,String}
+
+    excluded_connections = get_excluded_connections(connectivity)
+    for instruction in get_circuit_instructions(circuit)
+        qubits = get_connected_qubits(instruction)
+
+        for connection in excluded_connections
+            matching_indices = indexin(connection, qubits)
+            found_excluded_connection =
```

Comment:
this works as intended, even for 3 or more connected qubits, but it is a bit misleading in our case. If there is a 3 qubit gate, say Toffoli(1,2,3), it will reject connection (1,2) and (2,3) but not (1,3), which is making assumptions on the hardware implementation. Whats worse, for the forseeable future, 3 or more qubit gates wont be supported on the hardware anyway. I suggest an easier to understand implementation:

```
function are_gates_at_excluded_connections(
    connectivity::Union{LineConnectivity,LatticeConnectivity},
    circuit::QuantumCircuit,
)::Tuple{Bool,String}

    excluded_connections = get_excluded_connections(connectivity)
    for instruction in get_circuit_instructions(circuit)
        qubits = get_connected_qubits(instruction)

        if length(qubits) == 2
            candidate_connection = Tuple{Int64,Int64}(sort(qubits))
            if candidate_connection ∈ excluded_connections
                gate_name = typeof(get_gate_symbol(instruction))
                message =
                    "the $gate_name gate on qubits $qubits cannot be applied " *
                    "since connection $candidate_connection is unavailable"
                return (true, message)
            end
        end

        if length(qubits)>2
            throw(DomainError(are_gates_at_excluded_connections, "not implemented for gates with more that 2 connected qubits"))
        end

    end

    return (false, "")
end

```